tool/lib/commands/rollback.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io' as io;
+
+import 'package:args/command_runner.dart';
+
+import '../model.dart';
+
+class RollbackCommand extends Command {
+  RollbackCommand() : super() {
+    argParser.addOption('to-version');
+  }
+
+  @override
+  String get name => 'rollback';
+
+  @override
+  String get description => 'Rolls back to a specific DevTools version';
+
+  @override
+  Future run() async {
+    final repo = DevToolsRepo.getInstance();
+    print('DevTools repo at ${repo.repoPath}.');
+
+    final tempDir =
+        (await io.Directory.systemTemp.createTemp('devtools-rollback'))
+            .absolute;
+    print('file://${tempDir.path}');
+    final tarball = io.File(
+      '${tempDir.path}/devtools.tar.gz',
+    );
+    final extractDir =
+        await io.Directory('${tempDir.path}/extract/').absolute.create();
+    final client = io.HttpClient();
+    final version = argResults['to-version'];
+    print('downloading tarball to ${tarball.path}');
+    final tarballRequest = await client.getUrl(Uri.http(
+        'storage.googleapis.com',
+        'pub-packages/packages/devtools-$version.tar.gz'));
+    final tarballResponse = await tarballRequest.close();
+    await tarballResponse.pipe(tarball.openWrite());
+    print('Tarball written; unzipping.');
+
+    await io.Process.run(
+      'tar',
+      ['-x', '-z', '-f', tarball.path.split('/').last, '-C', extractDir.path],
+      workingDirectory: tempDir.path,
+    );
+    print('file://${tempDir.path}');
+
+    final buildDir = io.Directory('${repo.repoPath}/packages/devtools/build/');
+    await buildDir.delete(recursive: true);
+    await io.Directory('${extractDir.path}build/')
+        .rename('${repo.repoPath}/packages/devtools/build/');
+
+    print('Build outputs from Devtools version $version checked out and moved '
+        'to ${buildDir.path}');
+    print('To complete the rollback, go to ${repo.repoPath}/packages/devtools, '
+        'rev pubspec.yaml, update the changelog, unhide build/ from the '
+        'packages/devtools/.gitignore file, then run pub publish.');
```

Comment:
Is it also necessary to change the version in devtools.dart or devtools_app/pubspec.yaml?

Reply:
This approach will only make changes to package:devtools, not package:devtools_app.

Rather than fetch and rebuild the old app (which could introduce the same error if a dart2js or flutter bug caused the break), it fetches and re-distributes the old, working build.

As a result, changing any files that go into the distributed build won't have any effect.

So, if v0.1.15 was a rollback to v0.1.13, we would be shipping the v0.1.13 binary.  We'd want to reflect this in the changelog, with a format like:

```
# 0.1.15 (2020-02-06)
* Roll back to 0.1.13 from 0.1.14.  Reverts breakage caused by #1610
```

The devtools homepage which lists the app version would call itself version 0.1.13 in this example.

Reply:
Related to this concern, we should probably start logging exactly which version of Flutter/dart2js we build each release of devtools with. That way we would be able to reliably reproduce issues locally.